### PR TITLE
[in_app_purchase] Fix app exceptions caused by missing App Store receipt

### PR DESCRIPTION
**Pull request by renefloor**
## Description

Missing App Store receipt can happen for different scenarios, thus shouldn't trigger an exception.
Reboot of PR #2862

## Related Issues

https://github.com/flutter/flutter/issues/43957

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ x ] All existing and new tests are passing.
- [ x ] I updated/added relevant documentation (doc comments with `///`).
- [ x ] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [ x ] I read and followed the [Flutter Style Guide].
- [ x ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [ x ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ x ] I updated CHANGELOG.md to add a description of the change.
- [ x ] I signed the [CLA].
- [ x ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ x ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 1]
```diff
@@ -9,18 +9,22 @@ @implementation FIAPReceiptManager
 
 - (NSString *)retrieveReceiptWithError:(FlutterError **)error {
   NSURL *receiptURL = [[NSBundle mainBundle] appStoreReceiptURL];
-  NSData *receipt = [self getReceiptData:receiptURL];
+  NSError *err;
+  NSData *receipt = [self getReceiptData:receiptURL error:&err];
+  if (err) {
```
Comment (renefloor) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:14]: This needs to be tested

**Review by renefloor — COMMENTED** [review]
Should be good after xctested

File: packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/InAppPurchasePluginTests.m [hunk 2]
```diff
@@ -233,6 +234,25 @@ - (void)testRetrieveReceiptData {
...
+                     arguments:nil];
+  __block NSDictionary* result;
+  self.receiptManagerStub.returnError = YES;
+  [self.plugin handleMethodCall:call
+                         result:^(id r) {
+                           result = r;
+                           [expectation fulfill];
+                         }];
+  [self waitForExpectations:@[ expectation ] timeout:5];
+  NSLog(@"%@", result);
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/InAppPurchasePluginTests.m:253]: Please don't leave debug logging in tests.

File: packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/Stubs.h [hunk 3]
```diff
@@ -54,6 +54,7 @@ API_AVAILABLE(ios(11.2), macos(10.13.2))
 @end
 
 @interface FIAPReceiptManagerStub : FIAPReceiptManager
+@property(assign, nonatomic) BOOL returnError;
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/Stubs.h:57]: Needs a declaration comment. Test code is code that people need to understand and maintain like any other code, so has the same requirements on having comments explaining it.

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 4]
```diff
@@ -9,18 +9,22 @@ @implementation FIAPReceiptManager
 
 - (NSString *)retrieveReceiptWithError:(FlutterError **)error {
   NSURL *receiptURL = [[NSBundle mainBundle] appStoreReceiptURL];
-  NSData *receipt = [self getReceiptData:receiptURL];
+  NSError *err;
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:12]: Please don't distinguish two otherwise identical variables by abbreviating one and not the other; it doesn't do anything to help the reader understand. Call this one `receiptDataError` or something that explains how it differs from `error`.

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 5]
```diff
@@ -9,18 +9,22 @@ @implementation FIAPReceiptManager
...
+  NSData *receipt = [self getReceiptData:receiptURL error:&err];
+  if (err) {
+    *error = [FlutterError errorWithCode:[[NSString alloc] initWithFormat:@"%li", (long)err.code]
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:15]: ObjC convention is that error pointer pointer parameters can be null, so this should check before assigning.

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 6]
```diff
@@ -9,18 +9,22 @@ @implementation FIAPReceiptManager
...
+                                 message:err.domain
+                                 details:err.userInfo];
+    return nil;
+  }
   if (!receipt) {
-    *error = [FlutterError errorWithCode:@"storekit_no_receipt"
-                                 message:@"Cannot find receipt for the current main bundle."
-                                 details:nil];
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:16]: Why was this removed? There's now a codepath that returns nil but doesn't set the error.

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 7]
```diff
@@ -9,18 +9,22 @@ @implementation FIAPReceiptManager
...
     return nil;
   }
   return [receipt base64EncodedStringWithOptions:kNilOptions];
 }
 
-- (NSData *)getReceiptData:(NSURL *)url {
-  return [NSData dataWithContentsOfURL:url];
+- (NSData *)getReceiptData:(NSURL *)url error:(NSError **)error {
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:26]: Where is this method declared? If it's being overridden by a subclass for testing it should be declared in a private header.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by vytautas-pranskunas-** [comment]
any updates?

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m (hunk 6 above)
Comment (renefloor) [reply packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:16]: https://developer.apple.com/documentation/foundation/nsdata/1547238-datawithcontentsofurl

> This method may fail and return nil for various reasons, like bad URLs and file-not-found errors. In this case, the reason for error is indicated by errorPtr.

If this is nil it should always have given an error. This `if(!receipt)` can be removed completely, but that might give nullpointer if somehow the StoreKit does give nil without an error. I added an error, but I'm also fine with removing it completely.

**Comment by renefloor** [comment]
> any updates?

@vytautas-pranskunas- I was enjoying my weekend. The requested changes could wait till Monday ;)

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m (hunk 6 above)
Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:16]: Ah, then I would just simplify this down to:
`if (!receipt || receiptError) {...}`
with the first error construction code. If receiptError is nil it'll still work, just not give any extra info (which is fine for a case that is documented not to happen).

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m [hunk 8]
```diff
@@ -5,22 +5,41 @@
 #import "FIAPReceiptManager.h"
 #import <Flutter/Flutter.h>
 
+@interface FIAPReceiptManager ()
+
+- (NSData *)getReceiptData:(NSURL *)url error:(NSError **)error;
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPReceiptManager.m:10]: Needs a declaration comment (Primarily explaining what it does, but secondarily mentioning that it exists to be stubbed out in tests.)

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Comment by hiddenfloret** [comment]
I think this PR introduced a new bug, where app crashes when `buyNonConsumable` is called. See the above issue for details.
